hivelocity/instances_v2.go

```diff
@@ -0,0 +1,128 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package hivelocity
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/hivelocity/hivelocity-cloud-controller-manager/client"
+	"github.com/hivelocity/hivelocity-cloud-controller-manager/pkg/hvutils"
+	corev1 "k8s.io/api/core/v1"
+	cloudprovider "k8s.io/cloud-provider"
+)
+
+// HVInstancesV2 implements cloudprovider.InstanceV2.
+type HVInstancesV2 struct {
+	API client.API
+}
+
+var _ cloudprovider.InstancesV2 = &HVInstancesV2{}
+
+// GetHivelocityDeviceIDFromNode returns the deviceID from a Node.
+// Example: If Node.Spec.ProviderID is "hivelocity://123", then 123
+// will be returned.
+func GetHivelocityDeviceIDFromNode(node *corev1.Node) (int32, error) {
+	providerPrefix := providerName + "://"
+	if !strings.HasPrefix(node.Spec.ProviderID, providerPrefix) {
+		return 0, fmt.Errorf("missing prefix %q in node.Spec.ProviderID %q",
+			providerPrefix, node.Spec.ProviderID)
+	}
+	deviceID, err := strconv.ParseInt(node.Spec.ProviderID[len(providerPrefix):], 10, 32)
+	if err != nil {
+		return 0, fmt.Errorf("failed to convert node.Spec.ProviderID %q to int32",
+			node.Spec.ProviderID)
+	}
+	return int32(deviceID), nil
+}
+
+// InstanceExists returns true if the instance for the given node exists according to the cloud provider.
+// Use the node.name or node.spec.providerID field to find the node in the cloud provider.
+func (i2 *HVInstancesV2) InstanceExists(ctx context.Context, node *corev1.Node) (bool, error) {
+	deviceID, err := GetHivelocityDeviceIDFromNode(node)
+	if err != nil {
+		return false, fmt.Errorf("GetHivelocityDeviceIDFromNode(node) failed: %w", err)
+	}
+	_, err = i2.API.GetBareMetalDevice(ctx, deviceID)
+	if err == client.ErrNoSuchDevice {
+		return false, nil
+	}
+	if err != nil {
+		return false, fmt.Errorf("i2.API.GetBareMetalDeviceIdResource(deviceID) failed: %w", err)
```

Comment:
I changed fmt.Errorf() this. It looks like this now:

```
func (i2 *HVInstancesV2) InstanceExists(ctx context.Context, node *corev1.Node) (bool, error) {
	deviceID, err := getHivelocityDeviceIDFromNode(node)
	if err != nil {
		return false, fmt.Errorf("InstanceExists failed: %w", err)
	}
	_, err = i2.client.GetBareMetalDevice(ctx, deviceID)
	if errors.Is(err, client.ErrNoSuchDevice) {
		return false, nil
	}
	if err != nil {
		return false, fmt.Errorf("InstanceExists failed: %w", err)
	}
	return true, nil
}
```

Looking at it again, I think this could be simplified to this (just return the err)

```
func (i2 *HVInstancesV2) InstanceExists(ctx context.Context, node *corev1.Node) (bool, error) {
	deviceID, err := getHivelocityDeviceIDFromNode(node)
	if err != nil {
		return false, err
	}
	_, err = i2.client.GetBareMetalDevice(ctx, deviceID)
	if errors.Is(err, client.ErrNoSuchDevice) {
		return false, nil
	}
	if err != nil {
		return false, err
	}
	return true, nil
}
```

What do you think?

Reply:
If you have multiple places in the code where the same error would be returned, it can be helpful to have the error stack with `InstanceExists failed`. If this is not the case, we don't need it.

Reply:
ok, then I think returning the static error is enough here. Fixed in 351ae50